net_processing: make any misbehavior trigger immediate discouragement

sipa opened the pull request:
So far, discouragement of peers triggers when their misbehavior score exceeds 100 points. Most types of misbehavior increment the score by 100, triggering immediate discouragement, but some types do not. This PR makes all increments equal to either 100 (meaning any misbehavior will immediately cause disconnection and discouragement) or 0 (making the behavior effectively unconditionally allowed), and then removes the logic for score accumulation.

This simplifies the code a bit, but also makes protocol expectations clearer: if a peer misbehaves, they get disconnected. There is no good reason why certain types of protocol violations should be permitted 4 times (howmuch=20) or 9 times (howmuch=10), while many others are never allowed. Furthermore, the distinction between these looks arbitrary.


The specific types of misbehavior that are changed to 100 are:
* Sending us a `block` which does not connect to our header tree (which necessarily must have been unsollicited). [used to be score 10]
* Sending us a `headers` with a non-continuous headers sequence. [used to be score 20]
* Sending us more than 1000 addresses in a single `addr` or `addrv2` message [used to be score 20]
* Sending us more than 50000 invs in a single `inv` message [used to be score 20]
* Sending us more than 2000 headers in a single `headers` message [used to be score 20]

The specific types of misbehavior that are changed to 0 are:
* Sending us 10 (*) separate BIP130 headers announcements that do not connect to our block tree [used to be score 20]
* Sending us more than 8 headers in a single `headers` message (which thus does not get treated as a BIP130 announcement) that does not connect to our block tree. [used to be score 10]

I believe that none of these behaviors are unavoidable, except for the one marked (*) which can in theory happen still due to interaction between BIP130 and variations in system clocks (the max 2 hour in the future rule). This one has been removed entirely. In order to remove the impact of the bug it was designed to deal with, without relying on misbehavior, a separate improvement is included that makes `getheaders`-tracking more accurate.

In another unrelated improvement, this also gets rid of the 8 header limit heuristic to determine whether an incoming non-connecting `headers` is a potential BIP130 announcement, as this rule is no longer needed to prevent spurious Misbehavior. Instead, any non-connecting `headers` is now treated as a potential announcement.

sipa committed:
net_processing: make all Misbehaving increments = 100

This removes the need to actually track misbehavior score (see further commit), because any
Misbehaving node will immediately hit the discouragement threshold.

sipa committed:
net_processing: remove Misbehavior score and increments

This is now all unused.

instagibbs reviewed (commented):
concept ACK

going to run on a listening node with logging to test

1440000bytes commented:
Concept ACK

brunoerg commented:
Concept ACK

0xB10C commented:
I have been monitoring peer misbehavior with the misbehaving tracepoint from https://github.com/bitcoin/bitcoin/pull/25832. In the past weeks, I've seen the following misbehavior across 10 nodes:
- `header with invalid proof of work` [100]: very frequent
- `mutated block` [100]: these are frequent. I was running #29412 where we disconnect for each mutated block (compared to only disconnecting for mutated blocks we hadn't seen before)
- `invalid header received` [10]: these are infrequent ("Sending us more than 8 headers in a single headers message")
- `headers message size = X` [20]: only seen one occurrence ("Sending us more than 2000 headers in a single headers message that does not connect to our block tree")

I've looked at a few of these `invalid header received` misbehaviors and they all came from the same IP address with a (fake?) v26.0 Bitcoin Core user agent and always claiming their start height to be 153681 (in the version msg). Seems ok to disconnect these?

I'll keep an eye on further misbehavior.

Sjors commented:
> due to interaction between BIP130 and variations in system clocks (the max 2 hour in the future rule)

Perhaps we can tolerate this if the block is less than 10 hours in the future and disconnect if it's more? It seems cheap enough to process slightly-too-far-in-the-future headers, and they are enormously expensive to produce.

Sjors commented:
https://github.com/Sjors/bitcoin/commit/00a7dccc1c6957ce5a6a00d5322a25f949956e6d is one approach of dealing with the time difference issue. If any of the headers are `> MAX_FUTURE_BLOCK_TIME * 10` in the future we disconnect with `time-too-new`. We could perhaps increase the number to allow for even more incorrect clocks, but I'm not sure how useful that is.

sipa committed:
net_processing: drop Misbehavior for unconnecting headers

This misbehavior was originally intended to prevent bandwidth wastage due to
actually observed very broken (but likely non-malicious) nodes that respond
to GETHEADERS with a response unrelated to the request, triggering a request
cycle.

This has however largely been addressed by the previous commit, which causes
non-connecting HEADERS that are received while a GETHEADERS has not been
responded to, to be ignored, as long as they do not time out (2 minutes).
With that, the specific misbehavior is largely irrelevant (for inbound peers,
it is now harmless; for outbound peers, the eviction logic will eventually
kick them out if they're not keeping up with headers at all).

Sjors commented on src/net_processing.cpp:
```diff
@@ -1833,8 +1833,7 @@ bool PeerManagerImpl::MaybePunishNodeForBlock(NodeId nodeid, const BlockValidati
     case BlockValidationResult::BLOCK_MISSING_PREV:
-        // TODO: Handle this much more gracefully (10 DoS points is super arbitrary)
-        if (peer) Misbehaving(*peer, 10, message);
+        if (peer) Misbehaving(*peer, 100, message);
```
This is only triggered by `AcceptBlockHeader`, which typically happens _after_ `if(!headers_connect_blockindex)` causes an early return in `ProcessHeadersMessage`.

In other words, disconnecting here does not get in the way of `HandleUnconnectingHeaders` processing.

Presumably (I didn't check) this can be reached if we get an unsolicited BLOCK message. In which case it seems fine to disconnect.

It can also be reached when using the `submitblock(header)` RPC, but in that case there is no `peer`.

sipa committed:
net_processing: drop 8 headers threshold for incoming BIP130

With the Misbehavior score gone for non-connecting headers (see previous
commit), there is no need to only treat headers messages with up to 8
headers as potential BIP130 announcements. BIP130 does not specify such
a limit; it was purely a heuristic.

Sjors commented on src/net_processing.cpp:
```diff
@@ -2533,7 +2532,7 @@ bool PeerManagerImpl::CheckHeadersPoW(const std::vector<CBlockHeader>& headers,
     // Are these headers connected to each other?
     if (!CheckHeadersAreContinuous(headers)) {
-        Misbehaving(peer, 20, "non-continuous headers sequence");
+        Misbehaving(peer, 100, "non-continuous headers sequence");
```
I don't know why we ever tolerated this, but it's been that way since headers-first was added in #4468. Perhaps to avoid a network split in the case where headers-first had an implementation bug?

Note that the `getheaders` message is older, see https://github.com/bitcoin/bitcoin/pull/4468/commits/341735eb8f42e898cf9d4d130709471e5d01abe2#r17027597

Sjors commented on src/net_processing.cpp:
```diff
@@ -3782,7 +3781,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         if (vAddr.size() > MAX_ADDR_TO_SEND)
         {
-            Misbehaving(*peer, 20, strprintf("%s message size = %u", msg_type, vAddr.size()));
+            Misbehaving(*peer, 100, strprintf("%s message size = %u", msg_type, vAddr.size()));
```
I believe this limit was introduced a decade ago in #5419, at least on the sender side.

Sjors commented on src/net_processing.cpp:
```diff
@@ -3864,7 +3863,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         if (vInv.size() > MAX_INV_SZ)
         {
-            Misbehaving(*peer, 20, strprintf("inv message size = %u", vInv.size()));
+            Misbehaving(*peer, 100, strprintf("inv message size = %u", vInv.size()));
```
The maximum size of an `inv` message has be 50,000 bytes since Satoshi sneaked in that limit in e2c2648c14f4b87d331dbc30f0f2bd4aab9ce7e6.

Sjors commented on src/net_processing.cpp:
```diff
@@ -4651,7 +4650,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         unsigned int nCount = ReadCompactSize(vRecv);
         if (nCount > MAX_HEADERS_RESULTS) {
-            Misbehaving(*peer, 20, strprintf("headers message size = %u", nCount));
+            Misbehaving(*peer, 100, strprintf("headers message size = %u", nCount));
```
This limit has been in place since #4468

Sjors reviewed (commented):
I looked at each of the limits in aee3b67d43eea083d7299840dd11a61d4cf797b9 to confirm they're plenty-old enough to just disconnect anyone that breaks them.

Will look at the newly pushed commits later.

Sjors commented:
Reviewed dee6ca279cac97e167c82b0b2a61d2d2f8df1d47

I'll run the log commit for a while.

I'm still a bit on the fence about a903bba4d65720265659c6d4dae6cd8d9ec89c0f.

> With that, the specific misbehavior is largely irrelevant (for inbound peers, it is now mostly harmless; ...

Some context here for other reviewers, when you look at `SelectNodeToEvict` one of the criteria that prevents eviction is if they sent us novel blocks. This means that nodes that keep sending us useless headers, will probably not enjoy this protection and eventually get replaced. Unless they _also_ give use useful data, in which case they're ... useful - and we can just humor their headers shenaningans.

> ...  for outbound peers, the eviction logic will eventually kick them out if they're not keeping up with headers at all).

IIUC outbound peer eviction is handled in `PeerManagerImpl::ConsiderEviction`, which loses its patience after `CHAIN_SYNC_TIMEOUT + HEADERS_RESPONSE_TIME`. That seems a bit late though, if all our outbound peers are slow? Even if that's rare, it seems better to have as many "fit" outbound peers as possible.


Also... It might still be useful, even if not needed for security, to _know_ if a peer sent us disconnecting headers. Although the `NET` log still provides a log message `received header %s: missing prev block %s, sending getheaders (%d) to end (peer=%d)`, we don't know the outcome.

Tracking if the peer did this before, e.g. with something like `m_chicken_little`, would make this easier. But not sure if it's worth the complexity. It seems all this would do is reveal that someone tried an attack that we believe to be useless.

mzumsande commented on src/net_processing.cpp:
```diff
@@ -129,8 +129,6 @@ static constexpr double BLOCK_DOWNLOAD_TIMEOUT_BASE = 1;
 /** Maximum number of headers to announce when relaying blocks with headers message.*/
 static const unsigned int MAX_BLOCKS_TO_ANNOUNCE = 8;
-/** Maximum number of unconnecting headers announcements before DoS score */
-static const int MAX_NUM_UNCONNECTING_HEADERS_MSGS = 10;
```
The commit message of a903bba4d65720265659c6d4dae6cd8d9ec89c0f says that https://github.com/bitcoin/bitcoin/pull/25454 has largely addressed the problem of a request cycle, but I don't understand why:
If we receive an unconnecting header, we send a `GetHeaders`. But if our peer is broken in a way that they ignore the contents of our locator and send us unconnecting headers again, we clear `m_last_getheaders_timestamp`, call `ProcessHeadersMessage` and immediately send a `GetHeaders` again (we wold only delay that if `m_last_getheaders_timestamp` hadn't been cleared)  - and after this PR we wouldn't discourage the peer anymore so that this could go on forever. How did #25454 address that?

Sjors replied:
Glad you caught this @mzumsande.

mzumsande reviewed (commented):
Concept ACK, a few doc nits and one question below.

mzumsande commented on src/net_processing.cpp:
```diff
@@ -610,7 +610,7 @@ class PeerManagerImpl final : public PeerManager
     /** Deal with state tracking and headers sync for peers that send the
      * occasional non-connecting header (this can happen due to BIP 130 headers
```
maybe just drop the "occasional", but no big deal either way.

sr-gi reviewed (commented):
Concept ACK. I left some comments inline, plus some general ones that may be addressable with the change of functionality:

- Looks like `banmah.h` also needs updating, given there is no such thing as how much a peer misbehaves anymore:

  ```
  (line 37) // 2. Discouragement. If a peer misbehaves enough (see Misbehaving() in
  // net_processing.cpp)
  ```

- `p2p_filter.py` and `p2p_invalid_messages.py` are asserting logs looking for `Misbehaving`. Shouldn't it be better now to just check that the peer was disconnected in those cases? e.g:

  ```python
  self.log.info('Check that too large filter is rejected')
  with self.nodes[0].assert_debug_log(['Misbehaving']):
      filter_peer.send_and_ping(msg_filterload(data=b'\xbb'*(MAX_BLOOM_FILTER_SIZE+1)))
  ```

mzumsande commented on src/net_processing.cpp:
```diff
@@ -4529,7 +4488,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
                 ReadStatus status = partialBlock.InitData(cmpctblock, vExtraTxnForCompact);
                 if (status == READ_STATUS_INVALID) {
                     RemoveBlockRequest(pindex->GetBlockHash(), pfrom.GetId()); // Reset in-flight state in case Misbehaving does not result in a disconnect
```
yes, because some peers (e.g. ones with special permissions) don't get disconnected.

sipa committed:
net_processing: do not treat non-connecting headers as response

Since https://github.com/bitcoin/bitcoin/pull/25454 we keep track of the last
GETHEADERS request that was sent and wasn't responded to. So far, every incoming
HEADERS message is treated as a response to whatever GETHEADERS was last sent,
regardless of its contents.

This commit makes this tracking more accurate, by only treating HEADERS messages
which (1) are empty, (2) connect to our existing block header tree, or (3) are a
continuation of a low-work headers sync as responses that clear the "outstanding
GETHEADERS" state (m_last_getheaders_timestamp).

That means that HEADERS messages which do not satisfy any of the above criteria
will be ignored, not triggering a GETHEADERS, and potentially (for now, but see
later commit) increase misbehavior score.

Sjors commented:
I ran a slightly outdated version of this PR with the log commit e451338fa705a0a91b34f3fc53b8b921b1da9c63 for the last two weeks on a decently connected node. Only a single `MISBEHAVE` log entry:

```
168782-2024-03-07T15:58:10.608477Z Saw new header hash=00000000000000000002c88981de747ae237d80b74a5deb1a17943a76b296e3a height=833594 peer=1501 peeraddr=x.x.x.x:35630
168948-2024-03-07T15:58:10.947732Z UpdateTip: new best=00000000000000000002c88981de747ae237d80b74a5deb1a17943a76b296e3a height=833594 version=0x2319e000 log2_work=94.778914 tx=973821560 date='2024-03-07T15:57:26Z' progress=1.000000 cache=16.8MiB(112927txo)
169198:2024-03-07T15:58:19.055670Z MISBEHAVE: invalid header
```

This invalid header was sent 9 seconds after receiving a new valid block.

I'll continue to run with the latest version...

sr-gi commented on src/net_processing.cpp:
```diff
@@ -1848,7 +1835,7 @@ bool PeerManagerImpl::MaybePunishNodeForTx(NodeId nodeid, const TxValidationStat
     // The node is providing invalid data:
     case TxValidationResult::TX_CONSENSUS:
-        if (peer) Misbehaving(*peer, 100, "");
+        if (peer) Misbehaving(*peer, "");
```
I see. The message passed to `Misbehaving` ranges in how explicit it is though (from the actual reason to the category, without really specifying what, e.g. "invalid compact block"). That's why I was expecting this to at least signal the reason is due to a consensus discrepancy

sdaftuar commented:
Concept and code review ACK. I haven't done any testing (nor did I carefully review the tests), just read through the code and thought about the overall logical flow, which makes sense to me.

Sjors commented:
It's been two weeks and I don't see a single `MISBEHAVE` log entry.

mzumsande reviewed (commented):
Code Review ACK e7ccaa01e4e0ef502d5727367baa54bcdcf1fb83

willcl-ark commented:
Concept ACK.

IIUC this would not quite be enough to close down #9530 as it would not permit a header building on top of an invalid block with a valid header, as proposed there. Perhaps what is proposed here is superior though and ce238dd79bd3a1190a2229fac4502fa5742fc6e3 is enough to close that issue out if this goes in?

sr-gi commented on src/net_processing.cpp:
```diff
@@ -2628,6 +2628,8 @@ bool PeerManagerImpl::IsContinuationOfLowWorkHeadersSync(Peer& peer, CNode& pfro
     if (peer.m_headers_sync) {
         auto result = peer.m_headers_sync->ProcessNextHeaders(headers, headers.size() == MAX_HEADERS_RESULTS);
+        // If it is a valid continuation, we should treat the existing getheaders request as responded to.
+        if (result.success) peer.m_last_getheaders_timestamp = {};
```
in: ce238dd79bd3a1190a2229fac4502fa5742fc6e3

I think the comment 6 lines below is wrong now: 

```
// It should be impossible for the getheaders request to fail,
// because we should have cleared the last getheaders timestamp
// when processing the headers that triggered this call.
[...]
```

IIUC, `result.request_more` cannot be true if `result.success` is false (based on my understanding of [HeadersSyncState::ProcessNextHeaders](https://github.com/bitcoin/bitcoin/blob/07720b1cdd77399f32124641dbe1dd267eb0cf8b/src/headerssync.cpp#L66-L138)), therefore we will clear `m_last_getheaders_timestamp` right before entering the `request_more` scope.

I think this should also mean that there shouldn't be a way of bypassing this check (as the second part of the comment used to suggest), so the logging can be simplified

sr-gi reviewed (commented):
A couple of doc nits and a question below

sr-gi commented:
ACK [ae60d48](https://github.com/bitcoin/bitcoin/pull/29575/commits/ae60d485da33f238ed2186799da4e109d4edd3a1)

Something orthogonal to this, but that I realized working on the same part of the codebase and I wonder if it could be added here:

Looks like the result of `MaybePunishNodeForBlock` is never consumed in the codebase, plus the docs regarding what to return seem not to follow what the code is doing (if `peer` is not found, `Misbehaving` is not called, but the method returns `true`).

Should we just refactor the method to simply return `void`?

sipa committed:
net_processing: make MaybePunishNodeFor{Block,Tx} return void

sr-gi commented:
ACK [6eecba4](https://github.com/bitcoin/bitcoin/pull/29575/commits/6eecba475efd025eb011400af58621ad5823994e)

mzumsande reviewed (commented):
Code Review ACK 6eecba475efd025eb011400af58621ad5823994e

glozow reviewed (commented):
light code review / concept ACK 6eecba475efd025eb011400af58621ad5823994e

achow101 commented:
ACK 6eecba475efd025eb011400af58621ad5823994e

sr-gi commented on src/net_processing.cpp:
```diff
@@ -1828,8 +1828,7 @@ bool PeerManagerImpl::MaybePunishNodeForBlock(NodeId nodeid, const BlockValidati
         return true;
     // Conflicting (but not necessarily invalid) data or different policy:
```
I took a look at this to see if it would make sense to remove the comment, but looks like this is still consistent with the original approach, even after splitting the logic in two for txs and blocks:

https://github.com/bitcoin/bitcoin/commit/ef54b486d5333dfc85c56e6b933c81735196a25d

To me, it seems that, at the very least, this comment would need to be moved one case down, given `BlockValidationResult::BLOCK_MISSING_PREV` triggers misbehavior, so the comment doesn't apply there.
